### PR TITLE
[SPARK-21770][ML] ProbabilisticClassificationModel fix corner case: normalization of all-zero raw predictions

### PR DESCRIPTION
**Pull request by WeichenXu123**
## What changes were proposed in this pull request?

Fix probabilisticClassificationModel corner case: normalization of all-zero raw predictions, throw IllegalArgumentException with description.

## How was this patch tested?

Test case added.

### CODE REVIEW

File: mllib/src/main/scala/org/apache/spark/ml/classification/ProbabilisticClassifier.scala [hunk 1]
```diff
@@ -245,6 +245,13 @@ private[ml] object ProbabilisticClassificationModel {
         v.values(i) /= sum
         i += 1
       }
+    } else {
+      var i = 0
+      val size = v.size
+      while (i < size) {
+        v.values(i) = 1.0 / size
```
Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/ml/classification/ProbabilisticClassifier.scala]: Is Array.fill faster here?

File: mllib/src/test/scala/org/apache/spark/ml/classification/ProbabilisticClassifierSuite.scala [hunk 2]
```diff
@@ -23,6 +23,7 @@ import org.apache.spark.ml.param.ParamMap
 import org.apache.spark.ml.util.TestingUtils._
 import org.apache.spark.sql.{Dataset, Row}
 
+
```
Comment (heary-cao) [inline mllib/src/test/scala/org/apache/spark/ml/classification/ProbabilisticClassifierSuite.scala]: No need to be added line.

File: mllib/src/main/scala/org/apache/spark/ml/classification/ProbabilisticClassifier.scala (hunk 1 above)
Comment (WeichenXu123) [reply mllib/src/main/scala/org/apache/spark/ml/classification/ProbabilisticClassifier.scala]: But `Array.fill` will create a new array...

Comment (smurching) [reply mllib/src/main/scala/org/apache/spark/ml/classification/ProbabilisticClassifier.scala]: You could use `java.util.Arrays.fill` to update `v.values` in-place, but I'm not sure that it'll make a huge difference.

**Comment by smurching** [comment]
This looks good to me! @srowen would you be able to give it another look?

File: mllib/src/main/scala/org/apache/spark/ml/classification/ProbabilisticClassifier.scala [hunk 3]
```diff
@@ -245,6 +245,10 @@ private[ml] object ProbabilisticClassificationModel {
         v.values(i) /= sum
         i += 1
       }
+    } else {
+      var i = 0
+      val size = v.size
```
Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/ml/classification/ProbabilisticClassifier.scala]: What's the need for these two vars?

File: mllib/src/main/scala/org/apache/spark/ml/classification/ProbabilisticClassifier.scala [hunk 4]
```diff
@@ -245,6 +245,10 @@ private[ml] object ProbabilisticClassificationModel {
...
+      java.util.Arrays.fill(v.values, 1.0 / size)
```
Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/ml/classification/ProbabilisticClassifier.scala]: Import this, while you're at it.

**Review by srowen — CHANGES_REQUESTED** [review]

**Comment by sethah** [comment]
I'm confused how this issue was discovered in the first place. Did someone actually train an RF/DT and receive all zero probabilities? If so, shouldn't there be a unit test that recreates that scenario? 

Anyway, AFAICT the only way this could happen is if the RandomForest was trained on an empty DataFrame which couldn't happen because it would fail before the train method was called anyway.

**Comment by smurching** [comment]
@sethah I haven't heard of anybody hitting this issue in practice, but it did seem best to ensure that valid probability distributions would be produced regardless of input. There was some discussion of this in the JIRA: https://issues.apache.org/jira/browse/SPARK-21770

**Comment by sethah** [comment]
Ok, I guess I'm surprised that someone even noticed this...

So, basically, we are changing the behavior of a private function for a specific case which is actually impossible to ever run into. I don't see the need. I know @jkbradley mentioned that it might happen for linear models, but this method is only used for RandomForest and DecisionTree. @srowen mentioned that finding the root cause would be better, and seems to me the root cause is: there is no root cause. 

Anyway, if we are analyzing the behavior of this method without considering the context from which it's called, then we need to test for not only all zero probabilities but negative ones as well. Otherwise, we're shoring up one edge case without attending to others.

If it is not currently possible to call this method with all zero counts, and we don't want it to be possible ever in the future, why don't we just throw an error? I'm also fine with not changing this at all.

**Comment by srowen** [comment]
That's my sentiment too, that it is likely patching over another problem. I'd go with Joseph on this, but at least, issue a warning maybe?

I also would support throwing an exception to fail clearly and early in this case.

**Comment by jkbradley** [comment]
These are fair arguments.  I guess it makes sense to throw an exception; that's fine with me.

File: mllib/src/main/scala/org/apache/spark/ml/classification/ProbabilisticClassifier.scala [hunk 5]
```diff
@@ -245,6 +244,8 @@ private[ml] object ProbabilisticClassificationModel {
         v.values(i) /= sum
         i += 1
       }
+    } else {
```
Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/ml/classification/ProbabilisticClassifier.scala]: You can just `require(sum > 0)` at the top instead.

**Comment by WeichenXu123** [comment]
Jenkins, test this please.

**Comment by WeichenXu123** [comment]
ping @srowen Any other comments ?

File: mllib/src/main/scala/org/apache/spark/ml/classification/ProbabilisticClassifier.scala [hunk 6]
```diff
@@ -230,21 +230,19 @@ private[ml] object ProbabilisticClassificationModel {
    * Normalize a vector of raw predictions to be a multinomial probability vector, in place.
    *
    * The input raw predictions should be nonnegative.
-   * The output vector sums to 1, unless the input vector is all-0 (in which case the output is
-   * all-0 too).
+   * The output vector sums to 1, when the input vector is all-0, exception will be thrown.
```
Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/ml/classification/ProbabilisticClassifier.scala]: This isn't grammatical. You might just move the new note to a proper throws tag in the scaladoc here anyway

**Review by srowen — CHANGES_REQUESTED** [review]

File: mllib/src/main/scala/org/apache/spark/ml/classification/ProbabilisticClassifier.scala (hunk 6 above)
Comment (WeichenXu123) [reply mllib/src/main/scala/org/apache/spark/ml/classification/ProbabilisticClassifier.scala]: Updated. Thanks!

**Comment by WeichenXu123** [comment]
@srowen Any other comments? Thanks!

File: mllib/src/main/scala/org/apache/spark/ml/classification/ProbabilisticClassifier.scala [hunk 7]
```diff
@@ -230,21 +230,22 @@ private[ml] object ProbabilisticClassificationModel {
...
    *
    * The input raw predictions should be nonnegative.
-   * The output vector sums to 1, unless the input vector is all-0 (in which case the output is
-   * all-0 too).
+   * The output vector sums to 1.
    *
    * NOTE: This is NOT applicable to all models, only ones which effectively use class
    *       instance counts for raw predictions.
    */
+  @throws[IllegalArgumentException]("If the input vector is all-0 or including negative values")
```
Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/ml/classification/ProbabilisticClassifier.scala]: IllegalArgumentException is unchecked, so you don't need a `@throws` annotation even in Java. I was referring to a simple scaladoc tag.

File: mllib/src/main/scala/org/apache/spark/ml/classification/ProbabilisticClassifier.scala [hunk 8]
```diff
@@ -230,21 +230,22 @@ private[ml] object ProbabilisticClassificationModel {
...
+      "The input raw predictions should be nonnegative."))
     val sum = v.values.sum
-    if (sum != 0) {
-      var i = 0
-      val size = v.size
-      while (i < size) {
-        v.values(i) /= sum
-        i += 1
-      }
+    require(sum > 0, "All-0 vector is not allowed normalizing.")
```
Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/ml/classification/ProbabilisticClassifier.scala]: Text may be better as "Can't normalize the 0-vector" or similar

**Review by srowen — CHANGES_REQUESTED** [review]

**Review by srowen — APPROVED** [review]

**Comment by srowen** [comment]
Merged to master. I wasn't clear whether this was a pressing problem that needed to be backported.

**Pull request closed without merging**